### DISCUSSION TITLE
JSON or JSON lines

### DISCUSSION THREAD

**Discussion by dataMuley** [CMSgov/hospital-price-transparency#79] [Q&A] [answered]
Can the JSON file be JSON lines vs 'a single json object?'

**Comment by shaselton-usds** [comment]
Hey @dataMuley - Currently only single JSON files are allowed and JSONL is not. CSV is allowed as well in the appropriate formats that you can [find here](https://github.com/CMSgov/hospital-price-transparency/tree/master/documentation/CSV/templates) if you're interested in going that route.

**Reply by dataMuley** [reply]
JSONL is still listed as a .json file. If following the outlined schema, how does this not comply with such a rendering of JSON?

**Reply by shaselton-usds** [reply]
Multiple JSON root objects are not included in the defined schema as it wouldn't be valid JSON.

**Reply by dataMuley** [reply]
If 'root' objects still contain the same content, it doesn't change the included content nor is the schema different from 'line to line.' Due to the nature of the nested schema, the ingested content still has to be exploded to view 'readable' usable content which in theory means the 'root' content is repeated when shown in a 'table/dataframe', The lines approach in no way hinders that functionality but rather in some ways, aids in the ingestion of larger data sets.

**Reply by shaselton-usds** [reply]
JSONL is not considered valid JSON which is why it is technically a different format with a different extension. If you were to include JSONL in a .json file, it wouldn't be considered valid json due to having multiple root nodes. The extension for JSONL files is .jsonl.

You're correct that JSONL could be beneficial with larger data sets and if you find yourself running into some of those limitations maybe you can take advantage of streaming the JSON in chunks vs trying to load the whole set into memory. This is how the validator tool works.

**Reply by dataMuley** [reply]
Will CMS consider allowing a 'minor' deviation in the JSON layout in the future or now for a 'lines' approach? In the end, the desired schema can still be maintained yet efficiencies gained with such an approach. As you may/may not be aware, there are size constraints for 'cells' within various big data solutions. With the approach of a 'sole json object' these limits can be exceeded. 


https://kb.databricks.com/sql/cannot-grow-bufferholder-exceeds-size

**Reply by RidenourK** [reply, marked as answer]
@dataMuley

"CMS template" as defined [45 CFR 180.20](https://www.ecfr.gov/current/title-45/part-180#p-180.20(CMS%20template)) means a CSV format or JSON schema that CMS makes available for purposes of compliance with [§ 180.40(a)](https://www.ecfr.gov/current/title-45/section-180.40#p-180.40(a)). The CMS template is not currently available in JSONL. 

Thank you for you suggestion that we consider future Hospital Price Transparency – Data Dictionary GitHub repository updates to make the CMS template available in JSONL.
